File: library-core/src/main/java/com/mikepenz/iconics/utils/GenericsUtil.java
package com.mikepenz.iconics.utils;

import android.content.Context;
import android.text.TextUtils;

import java.lang.reflect.Field;
import java.util.ArrayList;

/**
 * Created by mikepenz on 03.08.15.
 */
public class GenericsUtil {

    /**
     * a helper to get the string fields from the R class
     *
     * @deprecated use {@link #getDefinedFonts(Context)} instead
     */
    @Deprecated
    public static String[] getFields(Context ctx) {
        return getDefinedFonts(ctx);
    }

    public static String[] getDefinedFonts(Context ctx) {
        Class rStringClass = resolveRClass(ctx.getPackageName());
        if (rStringClass != null) {
            return getDefinedFonts(ctx, rStringClass.getFields());
        }
        return new String[0];
    }

    public static String[] getDefinedProcessors(Context ctx) {
        Class rStringClass = resolveRClass(ctx.getPackageName());
        if (rStringClass != null) {
            return getDefinedProcessors(ctx, rStringClass.getFields());
        }
        return new String[0];
    }

    /**
     * a helper class to resolve the correct R Class for the package
     */
    private static Class resolveRClass(String packageName) {
        do {
            try {
                return Class.forName(packageName + ".R$string");
            } catch (ClassNotFoundException e) {
                packageName = packageName.contains(".") ? packageName.substring(0, packageName.lastIndexOf('.')) : "";
            }
        } while (!TextUtils.isEmpty(packageName));

        return null;
    }

    /**
     * A helper method to get a String[] out of a fieldArray
     *
     * @param fields R.strings.class.getFields()
     * @return a String[] with the string ids we need
     */
    private static String[] getDefinedFonts(Context ctx, Field[] fields) {
        ArrayList<String> fieldArray = new ArrayList<>();
        for (Field field : fields) {
            if (field.getName().contains("define_font_")) {
                fieldArray.add(getStringResourceByName(ctx, field.getName()));
            }
        }
        return fieldArray.toArray(new String[0]);
    }

    /**
     * A helper method to get a String[] out of a fieldArray
     *
     * @param fields R.strings.class.getFields()
     * @return a String[] with the string ids we need
     */
    private static String[] getDefinedProcessors(Context ctx, Field[] fields) {
        ArrayList<String> fieldArray = new ArrayList<>();
        for (Field field : fields) {
            if (field.getName().contains("define_processor_")) {
                fieldArray.add(getStringResourceByName(ctx, field.getName()));
            }
        }
        return fieldArray.toArray(new String[0]);
    }

    /**
     * helper class to retrieve a string by it's resource name
     */
    private static String getStringResourceByName(Context ctx, String resourceName) {
        String packageName = ctx.getPackageName();
        int resId = ctx.getResources().getIdentifier(resourceName, "string", packageName);
        if (resId == 0) {
            return "";
        } else {
            return ctx.getString(resId);
        }
    }
}

File: library-core/src/main/java/com/mikepenz/iconics/utils/GenericsUtil.kt
/*
 * Copyright 2019 Mike Penz
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.mikepenz.iconics.utils

import android.content.Context
import java.lang.reflect.Field

/**
 * Created by mikepenz on 03.08.15.
 */
object GenericsUtil {

    /** Helper to get the string fields with name starts "define_font_" from the R class */
    @JvmStatic fun getDefinedFonts(ctx: Context): Array<out String> {
        return resolveRClass(ctx.packageName)?.let { getDefinedFonts(ctx, it.fields) }.orEmpty()
    }

    /** Helper to get the string fields with name starts "define_processor_" from the R class */
    @JvmStatic fun getDefinedProcessors(ctx: Context): Array<out String> {
        return resolveRClass(ctx.packageName)?.let { getDefinedProcessors(ctx, it.fields) }
                .orEmpty()
    }

    /** Helper to resolve the correct R Class for the package */
    @JvmStatic private fun resolveRClass(packageName: String): Class<*>? {
        var tempPackageName = packageName
        do {
            try {
                return Class.forName("$tempPackageName.R\$string")
            } catch (e: ClassNotFoundException) {
                tempPackageName = tempPackageName.substringBeforeLast('.', "")
            }
        } while (tempPackageName.isNotBlank())

        return null
    }

    /**
     * Helper to get a array of strings out of a fieldArray
     *
     * @param fields R.strings.class.getFields()
     * @return a array of strings with the string ids we need
     */
    @JvmStatic private fun getDefinedFonts(ctx: Context, fields: Array<Field>): Array<String> {
        return fields.filter { it.name.contains("define_font_") }
                .map { getStringResourceByName(ctx, it.name) }
                .toTypedArray()
    }

    /**
     * A helper method to get a String[] out of a fieldArray
     *
     * @param fields R.strings.class.getFields()
     * @return a array of strings with the string ids we need
     */
    @JvmStatic private fun getDefinedProcessors(ctx: Context, fields: Array<Field>): Array<String> {
        return fields.filter { it.name.contains("define_processor_") }
                .map { getStringResourceByName(ctx, it.name) }
                .toTypedArray()
    }

    /** A helper method to retrieve a string by it's resource name */
    @JvmStatic private fun getStringResourceByName(ctx: Context, resourceName: String): String {
        val resId = ctx.resources.getIdentifier(resourceName, "string", ctx.packageName)
        return if (resId == 0) "" else ctx.getString(resId)
    }
}